Issue#64 library mapping

renannetto opened the pull request:
This adds LibraryMapping (to link netlist to standard cells) and PlacementMapping (to get information such as cell dimensions and pin positions, given entities from the netlist)

renannetto committed:
added LibraryMapping to circuit namespace.

renannetto committed:
added PlacementMapping to get cell geometry and pin location

renannetto committed:
added placement mapping and methods in placement to place pads

renannetto committed:
added documentation for LibraryMapping and PlacementMapping

csguth commented on test/placement/placement_test.cpp:
```diff
@@ -30,9 +47,37 @@ TEST_CASE_METHOD(NetlistFixture, "Placement: placing a cell", "[placement]") {
     REQUIRE(!(placedCell1Location == placedCell2Location));
 }
+
+TEST_CASE_METHOD(NetlistFixture, "Placement: placing an input pad", "[placement]") {
```
This code duplication doesn't smell good for me.

Why not just using the **old approach**?:
- Consider the position of a pin a Derived Property (Its derived because its obtained by summing the position of its parent cell and the library offset).
- A pad is a special `StandardCell::Pin` without owner `Cell` and its offset is equals to its actual position.
- A pad has a correspondent `Netlist::Pin` which should be mapped to the respective `StandardCell::Pin`. Since the `Netlist::Pin` doesn't have a owner `Netlist::Cell`, its position is just the offest, obtained from the library.

renannetto replied:
I don't like the code repetition either, but it's kinda weird to consider the pad offset as its position, because typically we read the pin offsets from a LEF library, and the pads are not even present in this library.

csguth replied:
I see. It does make sense. Although we need a smart mechanism for getting the location of a given pin in two cases:

1. the pin is related to a pad: return this property you just created.
2. the pin has an owner cell: return its offet + owner's location.

csguth commented on test/circuit/library_mapping_test.cpp:
```diff
@@ -0,0 +1,53 @@
+}
+
+TEST_CASE("LibraryMapping: setting the standard cell of a cell", "[library_mapping][cell]") {
+    NetlistAndStdCellFixture netlistAndStdCellFixture;
```
Please, check if the Netlist still has the Standard Cell mapping.

renannetto replied:
Sorry, I didn't understand what you mean by that.

csguth replied:
Nevermind.. I thought `circuit::Netlist` already had the Standard Cell Mapping.

csguth commented on test/geometry/scalingtest.cpp:
```diff
@@ -12,7 +12,7 @@ TEST_CASE("Geometry: scaling a Point", "[geometry][operations]") {
 
     Point scaledPoint;
-    scale(pointFixture.point, scalingPoint, scaledPoint);
+    scale(pointFixture.point, scalingPoint.x(), scalingPoint.y(), scaledPoint);
```
I'd rather use:
`Point scaled = scale(pointFixture.point, scalingPoint);`
or 
`const Point scaled = scale(pointFixture.point, scalingPoint);`
or
`Point scaled = pointFixture.point * scalingPoint;`
or
`const Point scaled = pointFixture.point * scalingPoint;`

An even better interface would be: `auto scaled = pointFixture.point.scale(scalingPoint);`

renannetto replied:
I agree that those options are better, but I had problems when using something from type Location as scalingPoint, because the Location coordinates are micrometers, instead of doubles. It would be great if someone could fix that.

renannetto replied:
Actually, I think this could be fixed by adding another template to the Point type in the geometric operations, I'll try that.

csguth replied:
It sounds like a plan.

renannetto committed:
Changing operations interface to receive point.

renannetto committed:
Merge branch 'development' into issue#64-LibraryMapping

renannetto commented:
@csguth I updated the translate and scale methods signature to use one of those you suggested. The only problem was that it is not working with micron (or dbu) values, it must be a double value. So, I added a toPoint method in the Location type to transform it into a Point<double>. This way, we can use the geometric operations easily. Please review it so that we can finally close this pull request.
